http/src/main.rs

```diff
@@ -185,17 +190,29 @@ fn main() {
 
 fn serve<Types: IpfsTypes>(
     ipfs: &Ipfs<Types>,
+    listening_addr: Multiaddr,
 ) -> (std::net::SocketAddr, impl std::future::Future<Output = ()>) {
     use tokio::stream::StreamExt;
     use warp::Filter;
+
     let (shutdown_tx, mut shutdown_rx) = tokio::sync::mpsc::channel::<()>(1);
 
     let routes = v0::routes(ipfs, shutdown_tx);
     let routes = routes.with(warp::log(env!("CARGO_PKG_NAME")));
 
     let ipfs = ipfs.clone();
 
-    warp::serve(routes).bind_with_graceful_shutdown(([127, 0, 0, 1], 0), async move {
+    let components = listening_addr.iter().collect::<Vec<_>>();
+
+    use std::net::SocketAddr;
+    let socket_addr =
+        if let (Protocol::Ip4(ip), Protocol::Tcp(port)) = (&components[0], &components[1]) {
```

Comment:
I think this needs an additional check that the length of components is 2... Now, how to realize that.

My simplest solution (which doesnt turn out simple at all) is always to "just" call iter.next() three times, but that gets complicated quite fast in order to honor the "cannot call `iter.next()` after it has returned None" agreement.

Perhaps [slice pattern matching](https://doc.rust-lang.org/edition-guide/rust-2018/slice-patterns.html) could work here together with the vec collecting?